### PR TITLE
:warning: support delete validation in validator interface

### PR DESCRIPTION
**Pull request by awesomenix**
- Currently validator interface supports create and update
- Add support for delete as well, which could be a breaking change
- Users need to implement ValidateDelete since interface is updated

### CODE REVIEW

**Comment by mengqiy** [comment]
@awesomenix I'd like to hear your use case more before we decide to make change to the interface.

**Comment by mengqiy** [comment]
For breaking change, we should use `:warning:` I have updated the title.

**Comment by awesomenix** [comment]
@mengqiy In our case we use annotations to validate deletes on certain resource so that none of the deletes are accidental. We used Validating Webhooks which support validating DELETE operations as well (this was easy to override in kubebuilder v1), hence requesting the same in v2 as well.

BTW love the easy scaffolding of webhooks v2, way less code than the original

**Comment by mengqiy** [comment]
I'm thinking if we should support CONNECT operation. Trying to find if there is a common use case for it.

**Comment by awesomenix** [comment]
was thinking if we should add support for CONNECT as well, but wasnt sure of any use case. Honestly i do not know when it will be used, during resource get?

**Comment by mengqiy** [comment]
@awesomenix 
Re. DELETE operation, I saw this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#plumbing-existing-objects-to-delete-admission).

It seems upstream doesn't support it yet. No objects will be in the `AdmissionReview`.

File: pkg/webhook/admission/validator.go [hunk 1]
```diff
@@ -89,5 +90,17 @@ func (h *validatingHandler) Handle(ctx context.Context, req Request) Response {
 		}
 	}
 
+	if req.Operation == v1beta1.Delete {
+		err := h.decoder.Decode(req, obj)
```
Comment (mengqiy) [inline pkg/webhook/admission/validator.go]: It should be `err := h.decoder.DecodeRaw(req.OldObject, oldObj)` here

**Comment by awesomenix** [comment]
> KEP

Thanks for the reference. I see that 1.15 has the fix https://github.com/kubernetes/kubernetes/pull/76346.

Updating PR to reflect the same.

File: pkg/builder/webhook_test.go [hunk 2]
```diff
@@ -269,6 +269,104 @@ var _ = Describe("application", func() {
...
+    },
+    "namespace":"default",
+    "operation":"DELETE",
+    "object":null,
+    "oldObject":{
+      "replica":0
+    }
+  }
+}`)
+			By("sending a request to a mutating webhook path with correct request")
```
Comment (mengqiy) [inline pkg/builder/webhook_test.go]: validating

File: pkg/builder/webhook_test.go [hunk 3]
```diff
@@ -269,6 +269,104 @@ var _ = Describe("application", func() {
...
+			path = generateValidatePath(testValidatorGVK)
+			req = httptest.NewRequest("POST", "http://svc-name.svc-ns.svc"+path, reader)
+			req.Header.Add(http.CanonicalHeaderKey("Content-Type"), "application/json")
+			w = httptest.NewRecorder()
+			svr.WebhookMux.ServeHTTP(w, req)
+			Expect(w.Code).To(Equal(http.StatusOK))
+			Expect(w.Code).To(Equal(http.StatusOK))
```
Comment (mengqiy) [inline pkg/builder/webhook_test.go]: Duplicate line

**Comment by awesomenix** [comment]
@mengqiy thanks for the feedback. Updated PR.

File: pkg/webhook/admission/validator.go [hunk 4]
```diff
@@ -89,5 +90,19 @@ func (h *validatingHandler) Handle(ctx context.Context, req Request) Response {
 		}
 	}
 
+	if req.Operation == v1beta1.Delete {
+		// In reference to PR: https://github.com/kubernetes/kubernetes/pull/76346
+		// OldObject contains the object being deleted
+		err := h.decoder.DecodeRaw(req.OldObject, obj)
```
Comment (mengqiy) [inline pkg/webhook/admission/validator.go]: `req.OldObject` is an empty object when kube-apiserver is v1.14 or earlier.
`obj` will be a object with zero values everywhere.
`DecodeRaw` really should return an error when there is an empty object.
I created #529. We should merge that first.

**Pull request merged**